msm/src/mvlookup.rs

```diff
@@ -0,0 +1,171 @@
+//! Implement the protocol MVLookup <https://eprint.iacr.org/2022/1530.pdf>
+
+use std::iter;
+
+use ark_ff::{FftField, Field};
+use kimchi::circuits::domains::EvaluationDomains;
+use rand::{seq::SliceRandom, thread_rng, Rng};
+
+// TODO: Add more built-in lookup tables
+#[derive(Copy, Clone, Debug)]
+pub enum LookupTable {
+    RangeCheck16,
+    /// Custom lookup table
+    /// The index of the table is used as the ID, padded with the number of
+    /// built-in tables.
+    Custom(usize),
+}
+
+impl LookupTable {
+    /// Assign a unique ID to the lookup tables.
+    pub fn into_field<F: Field>(self) -> F {
+        match self {
+            LookupTable::RangeCheck16 => F::one(),
+            LookupTable::Custom(id) => F::from(id as u64) + F::one(),
+        }
+    }
+}
+
+/// Generic structure to represent a (vector) lookup the table with ID
+/// `table_id`.
+/// The structure represents the individual fraction of the sum described in the
+/// MVLookup protocol (for instance Eq. 8).
+/// The table ID is added to the random linear combination formed with the
+/// values. The combiner for the random linear combination is coined during the
+/// proving phase by the prover.
+#[derive(Debug, Clone)]
+pub struct Lookup<F> {
+    #[allow(dead_code)]
+    pub(crate) table_id: LookupTable,
+    #[allow(dead_code)]
+    pub(crate) numerator: F,
+    #[allow(dead_code)]
+    pub(crate) value: Vec<F>,
+}
+
+/// Represents a witness of one instance of the lookup argument
+#[derive(Debug)]
+pub struct LookupWitness<F> {
+    /// A list of functions/looked-up values.
+    #[allow(dead_code)]
+    pub(crate) f: Vec<Vec<Lookup<F>>>,
+    /// The table the lookup is performed on.
+    #[allow(dead_code)]
+    pub(crate) t: Vec<Lookup<F>>,
+    /// The multiplicity polynomial
+    #[allow(dead_code)]
+    pub(crate) m: Vec<F>,
+}
+
+// This should be used only for testing purposes.
+// It is not only in the test API because it is used at the moment in the
+// main.rs. It should be moved to the test API when main.rs is replaced with
```

Comment:
Agree, good to have it though. Maybe could be a `Random` instance instead of just an `impl` method.

Reply:
Opened follow-up https://github.com/o1-labs/proof-systems/issues/1825